### DISCUSSION TITLE
I2C device availability check

### DISCUSSION THREAD

**Discussion by andrico21** [arendst/Tasmota#19027] [Q&A]
Hi.

**Short question**: is there a way to detect I2C device availability/presence programmatically in Tasmota? Looking for a proper trigger to execute '_i2cdriver76 0_' and '_i2cdriver76 1_' commands via rule. Tasmota 12.5.0.0

**More details and context**.
Have a solar powered (daylight only) ESP32-C3 device with Sensirion SEN55 connected via I2C: it starts when sun is rising and turns off at sundown.

**Problem**: sometimes at dawn it seems that no enough juice coming from solar panel - controller starts, but there's no enough power for sensor and controller doesn't detect it. If I reboot controller via Web a few minutes later or just run _i2cdriver76 0_' and '_i2cdriver76 1_' commands - everything is fine.

'_i2cscan_' output while sensor is unavailable:
![image](https://github.com/arendst/Tasmota/assets/44984087/184eb5eb-17d1-4999-a70f-9b32611e258b)

**Comment by barbudor** [comment]
Command 'i2cscan' will tell you at which address there are i2c devices

**Reply by andrico21** [reply]
I've attached screenshot :) i2scan doesn't work, but switching the driver off and on works fine

**Reply by barbudor** [reply]
I2Cscan works fine as it tells you there is a problem
Up to you to handle that error message

**Comment by barbudor** [comment]
But I don't think that will be the best approach

You know that your device is attached. I2cdriver doesn't help you per se. What helps is the restart implied by those
So you'd better use rules to manage a timeout if you don't get the value for the sensor

**Reply by andrico21** [reply]
> I2cdriver doesn't help you per se.

but it actually helps as mentioned above - sensor becomes available. Sure I can make some kind of watchdog which will reboot a controller if there no values in MQTT channel for a while, but self-recovering logic inside of ESP seems more predictive

**Reply by barbudor** [reply]
Per design, Tasmota scan for I2cdevices at boot only
If your i2c chip doesn't respond, the recovery is to restart
Only you knows that such device is attached, not tasmota, and can take appropriate decision

I2cdriver command is not there to tell tasmota which devices are actually installed, just which devices that are possibly there.

**Comment by andrico21** [comment]
> Per design, Tasmota scan for I2cdevices at boot only If your i2c chip doesn't respond, the recovery is to restart Only you knows that such device is attached, not tasmota, and can take appropriate decision

I clearly understand when it scans for new devices: no any offence bud.
Hardware installation is static/permanent - device is always connected physically, but sometimes we have let's say to "reset the bus/device/whatever" or something like that.
Okay, so if I understand you right - logic might be following:
trigger some rule like on System#Boot and execute 'I2CScan' - but how to check its result in Tasmota? cannot find an example [here](https://tasmota.github.io/docs/Rules/), see the mention of '$|' operator, but no example how to check command output.
is the '`ON I2CScan#Data$|"Error 5" DO`' is proper approach?

**Reply by sfromis** [reply]
The syntax for rule triggers does not include string delimiters, and thus not spaces.

As the command I2CDriver will also trigger a restart, just doing the restart may be a more straightforward plan. A more general strategy would be to not try to detect the error, but instead detect the sensor working, as in providing data. If no data arrives after boot, then you could trigger a restart. What I mean is to setup a ruletimer at boot, and reset it when you get data. If it pops without getting reset, restart.

**Comment by andrico21** [comment]
So just to leave as artifact here: it seems additional 1sec delay for SEN5x init function resolved a problem.
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_103_sen5x.ino#L87

just added "delay(1000);" right before 'if (!Settings->flag6.sen5x_passive_mode) {' - now sensor is available after reboot in 100% of cases

**Reply by sfromis** [reply]
You probably need no source modification for that, using `SetOption46` you can have a power-on delay up to 2.55 seconds.

**Reply by andrico21** [reply]
unfortunately that didn't help - just reflashed to stock tasmota32c3.bin 13.4.0. No data coming from sensor without delay mentioned by me above:
```
20:30:24.954 MQT: tasmota/discovery/7CDFA1A3BAD0/sensors = {"sn":{"Time":"2024-04-14T20:30:24","ESP32":{"Temperature":43.0},"TempUnit":"C"},"ver":1} (retained)
20:30:34.214 WIF: Checking connection...
20:30:42.791 HTP: Console
20:30:50.487 CMD: i2cscan
20:30:50.488 SRC: WebConsole from 192.168.x.y
20:30:50.490 CMD: Grp 0, Cmd 'I2CSCAN', Idx 1, Len 0, Pld -99, Data ''
20:30:50.510 MQT: stat/tasmota_A3BAD0/RESULT = {"I2CScan":"Device(s) found on bus1 at 0x69"}
20:30:54.198 WIF: Checking connection...
20:31:00.912 HTP: Main Menu
20:31:01.381 CMD: setoption46
20:31:01.383 SRC: WebConsole from 192.168.x.y
20:31:01.384 CMD: Grp 0, Cmd 'SETOPTION', Idx 46, Len 0, Pld -99, Data ''
20:31:01.387 MQT: stat/tasmota_A3BAD0/RESULT = {"SetOption46":255}
20:31:14.216 WIF: Checking connection...
```

after a lot of reboots sensor becomes visible

**Comment by machdyne** [comment]
I was also having problems detecting a SEN55 and the 1 second delay appears to have solved the problem. Also tried SetOption46 with no luck.
